src/Adapter/EntityTranslation/EntityTranslator.php

```diff
@@ -0,0 +1,245 @@
+     */
+    protected $dbPrefix;
+
+    /**
+     * @param Db $db
+     * @param string $dbPrefix
+     * @param TranslatorInterface $translator
+     * @param DataLangCore $dataLang
+     */
+    public function __construct(
+        Db $db,
+        string $dbPrefix,
+        TranslatorInterface $translator,
+        DataLangCore $dataLang
+    ) {
+        $this->dataLang = $dataLang;
+        $this->db = $db;
+        $this->dbPrefix = $dbPrefix;
+        $this->translator = $translator;
+        $this->tableName = $this->buildTableNameFromDataLang($dataLang);
+    }
+
+    /**
+     * Translate the entity's data in database using reverse translation technique
+     *
+     * @param int $languageId
+     * @param int $shopId
+     *
+     * @throws LanguageNotFoundException
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    public function translate(int $languageId, int $shopId): void
+    {
+        $lang = new Language($languageId);
+        if (empty($lang->id)) {
+            throw new LanguageNotFoundException(new LanguageId($languageId));
+        }
+
+        $tableNameSql = bqSQL($this->tableName);
+
+        $shopWhere = ($this->shopFieldExists($tableNameSql))
+            ? sprintf(' AND `id_shop` = %d', $this->shopId)
+            : '';
+
+        // get table data
+        $sql = "SELECT * FROM `$tableNameSql` WHERE `id_lang` = $languageId"
+            . $shopWhere;
+
+        $tableData = $this->db->executeS($sql, true, false);
+
+        if (empty($tableData)) {
+            return;
+        }
+
+        $keys = $this->dataLang->getKeys();
+        $fieldsToUpdate = $this->dataLang->getFieldsToUpdate();
+
+        foreach ($tableData as $data) {
+            $updateWhere = [];
+            $updateFields = [];
+
+            // Construct update where
+            foreach ($keys as $key) {
+                $updateWhere[] = '`' . bqSQL($key) . '` = "' . pSQL($data[$key]) . '"';
+            }
+
+            // Construct update field
+            foreach ($fieldsToUpdate as $fieldName) {
+                if ('url_rewrite' === $fieldName && Language::$locale_crowdin_lang === $lang->locale) {
+                    continue;
+                }
+
+                $translatedField = $this->doTranslate($data, $fieldName);
+
+                if (!empty($translatedField) && $translatedField != $data[$fieldName]) {
+                    $updateFields[] = '`' . bqSQL($fieldName) . '` = "' . pSQL($translatedField) . '"';
+                }
+            }
+
+            // Update table
+            if (!empty($updateWhere) && !empty($updateFields)) {
+                $updateWhere = implode(' AND ', $updateWhere);
+                $updateFields = implode(', ', $updateFields);
+
+                $sql = "UPDATE `$tableNameSql`
+                    SET $updateFields
+                    WHERE $updateWhere AND `id_lang` = $languageId"
+                    . $shopWhere
+                    . ' LIMIT 1';
```

Comment:
```suggestion
                $sql = 'UPDATE `' . $tableNameSql . '` '
                    'SET $updateFields ' .
                    'WHERE ' . $updateWhere . ' AND `id_lang` = ' . $languageId . ' '
                    . $shopWhere
                    . ' LIMIT 1';
```

Reply:
There is an error in your suggestion. IMO I find the original more readable.

Reply:
I just don't like mixing ways to concatenate variables in strings :roll_eyes: